### PR TITLE
support dynamic IP address with DHCP

### PR DESCRIPTION
**Pull request by yaocw2020**
**issue**:  https://github.com/rancher/harvester-network-controller/issues/14
**solution**：
- Implement DHCP client for bridge.
- Start a DHCP daemon when create a bridge.
- Call os command `connmanctl` to configure DHCP for physical nic.

### CODE REVIEW

File: pkg/dhcp/client.go [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+				// rebind
+				lease, err = c.rebind(lease)
+				if err == nil {
+					klog.Infof("rebind, lease: %+v", lease)
+					t1.Reset(t1Timeout)
+					klog.Info("t1 reset")
+				} else {
+					klog.Errorf("rebind failed, error: %s", err.Error())
+					t1.Stop()
+					t2.Stop()
```
Comment (gitlawr) [inline pkg/dhcp/client.go:98]: `defer t1/2.Stop()` before the loop

File: pkg/bridge/bridge.go [hunk 2]
```diff
@@ -100,8 +127,14 @@ func (br *Bridge) BorrowAddr(lender *Link, vidList []uint16) error {
 		}
 	}
 
-	// transfer address
-	return transferAddr(lender, &Link{Link: br.Bridge})
+	// stop dhcp client for physical nic
```
Comment (guangbochen) [inline pkg/bridge/bridge.go:130]: check and skip DHCP client for physical NIC with static IP config.

File: pkg/bridge/bridge.go [hunk 3]
```diff
@@ -6,19 +6,46 @@ import (
...
+		DHCPSignal: make(chan bool),
+	}
+
+	go br.StartDHCPDaemon()
+
+	return br
+}
+
+// StartDHCPDaemon maintain dynamic ip address of bridge with dhcp
+func (br *Bridge) StartDHCPDaemon() {
```
Comment (guangbochen) [inline pkg/bridge/bridge.go:34]: prefer to call it `StartDHCPDaemonClient`

File: pkg/bridge/bridge.go [hunk 4]
```diff
@@ -6,19 +6,46 @@ import (
 
 	"github.com/vishvananda/netlink"
 	"k8s.io/klog"
+
+	"github.com/rancher/harvester-network-controller/pkg/dhcp"
```
Comment (guangbochen) [inline pkg/bridge/bridge.go:10]: format the imported library properly

File: pkg/dhcp/client_test.go [hunk 5]
```diff
@@ -0,0 +1 @@
+package dhcp
```
Comment (guangbochen) [inline pkg/dhcp/client_test.go:1]: remove this file if this is empty

File: pkg/dhcp/connman_dhcp.go [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+	"strings"
+
+	"github.com/vishvananda/netlink"
+)
+
+// Networking is powered by connman in k3os.
+// Thus, we use connmanctl to configure dhcp for physical nic
+
+const (
+	configDHCPCmd = "connmanctl config %s --ipv4 %s"
```
Comment (guangbochen) [inline pkg/dhcp/connman_dhcp.go:15]: add `commanctl` to the container deployment

File: pkg/dhcp/connman_dhcp.go [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+// ConnmanDHCP config dhcp with connmanctl
+func ConnmanDHCP(iface string, enable bool) error {
+	// connmanctl config <service> --ipv4 off
+	l, err := netlink.LinkByName(iface)
+	if err != nil {
+		return fmt.Errorf("netlink by name failed, error: %w, iface: %s", err, iface)
```
Comment (guangbochen) [inline pkg/dhcp/connman_dhcp.go:23]: text changes: `netlink name %s is invalid, error:...`

File: pkg/dhcp/connman_dhcp.go [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+	}
+
+	var op string
+	if enable {
+		op = "dhcp"
+	} else {
+		op = "off"
+	}
```
Comment (guangbochen) [inline pkg/dhcp/connman_dhcp.go:31]: using type string instead

File: pkg/dhcp/connman_dhcp.go [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+// Networking is powered by connman in k3os.
+// Thus, we use connmanctl to configure dhcp for physical nic
+
+const (
+	configDHCPCmd = "connmanctl config %s --ipv4 %s"
+)
+
+// ConnmanDHCP config dhcp with connmanctl
+func ConnmanDHCP(iface string, enable bool) error {
```
Comment (guangbochen) [inline pkg/dhcp/connman_dhcp.go:19]: prefer to use `ConfigConnman`

File: pkg/dhcp/client.go [hunk 10]
```diff
@@ -0,0 +1,150 @@
...
+	if err := c.Close(); err != nil {
+		klog.Errorf("close dhcp client failed, error: %s", err.Error())
+	}
+}
+
+// Start state-transition process of dhcp client
+func (c *Client) Start() {
+	for {
+		// DHCP State-transition: INIT --> BOUND
+		lease, err := c.Request(context.TODO(), dhcpv4.WithLeaseTime(defaultLeaseTime))
```
[the remark this answers is not in the document]
Comment (yaocw2020) [reply pkg/dhcp/client.go:54]: It's just for testing. I will modify it as 86400s before merging.

File: pkg/bridge/bridge.go (hunk 3 above)
Comment (yaocw2020) [reply pkg/bridge/bridge.go:34]: may be `StartDHCPClientDaemon` is better

File: pkg/dhcp/client.go [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+	"time"
+
+	"github.com/insomniacslk/dhcp/dhcpv4"
+	"github.com/insomniacslk/dhcp/dhcpv4/nclient4"
+	"github.com/insomniacslk/dhcp/netboot"
+	"k8s.io/klog"
+)
+
+// TODO for test
+const defaultLeaseTime = 16    // unit is second
```
Comment (gitlawr) [inline pkg/dhcp/client.go:17]: s/defaultLeaseTime/defaultLeaseSeconds/

File: pkg/dhcp/client.go [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+const defaultRetryTimes = 10
+const defaultRetryInterval = 2 * time.Minute
```
Comment (gitlawr) [inline pkg/dhcp/client.go:19]: retry in a couple seconds would be more reasonable

File: pkg/dhcp/client.go [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+		t2Timeout := lease.ACK.IPAddressLeaseTime(0) / 8 * 7
+		klog.Infof("t1: %s, t2: %s", t1Timeout, t2Timeout)
+		t1, t2 := time.NewTicker(t1Timeout), time.NewTicker(t2Timeout)
+		defer t1.Stop()
+		defer t2.Stop()
+
+		for {
+			select {
+			case <-t1.C:
+				klog.Info("t1 expired")
```
Comment (gitlawr) [inline pkg/dhcp/client.go:84]: remove this

File: pkg/bridge/bridge.go [hunk 14]
```diff
@@ -122,14 +169,26 @@ func (br *Bridge) ReturnAddr(returnee *Link, vidList []uint16) error {
...
+	}
+
+	if mode == dhcp.DHCP {
+		// stop bridge dhcp client
+		br.DHCPSignal <- false
+		// start dhcp client for physical nic
+		if err := returnee.startDHCP(); err != nil {
+			return fmt.Errorf("start dhcp client for link %s failed, error: %w", br.Name, err)
+		}
 	}
```
Comment (gitlawr) [inline pkg/bridge/bridge.go:188]: We are trying to return address to the NIC. Seems that the NIC is in `off` mode(not dhcp) at that moment.

**Review by guangbochen — APPROVED** [review]

**Pull request merged**